op-plasma/dastate.go

```diff
@@ -75,11 +53,30 @@ func NewState(log log.Logger, m Metricer) *State {
 }
 
 // IsTracking returns whether we currently have a commitment for the given key.
+// if the block number is mismatched we return false to ignore the challenge.
 func (s *State) IsTracking(key []byte, bn uint64) bool {
 	if c, ok := s.commsByKey[string(key)]; ok {
 		return c.blockNumber == bn
 	}
-	return false
+	// track the commitment knowing we may be in detached head and not have seen
+	// the commitment in the inbox yet.
+	s.TrackDetachedCommitment(key, bn)
+	return true
+}
+
+// TrackDetachedCommitment is used for indexing challenges for commitments that have not yet
+// been derived due to the derivation pipeline being stalled pending a commitment to be challenged.
+// Memory usage is bound to L1 block space during the DA windows, so it is hard and expensive to spam.
+// We do not append it yet in order to preserve the order.
+func (s *State) TrackDetachedCommitment(key []byte, bn uint64) {
+	c := &Commitment{
+		key:         key,
+		expiresAt:   bn,
+		blockNumber: bn,
+		canonical:   false,
+	}
+	s.log.Debug("tracking detached commitment", "blockNumber", c.blockNumber, "commitment", fmt.Sprintf("%x", key))
+	s.commsByKey[string(key)] = c
```

Comment:
Because we don't put this into a queue, we're not properly pruning commitments that come in through challenges.